cmd/src/dump.rs

```diff
@@ -0,0 +1,101 @@
+use engine_rocks::raw_util::CFOptions;
+use engine_rocks::{
+    self,
+    raw::{DBOptions, Env},
+    RocksEngine,
+};
+use engine_traits::{Iterable, RaftEngine, CF_DEFAULT, CF_RAFT};
+use kvproto::raft_serverpb::RaftLocalState;
+use protobuf::Message;
+use raft::eraftpb::Entry;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+pub type RaftDumpData = HashMap<u64, (Vec<Entry>, RaftLocalState)>;
+
+/// Check the potential original raftdb directory and try to dump data out.
+///
+/// TiKV only looks for the name in the directory where `raft-engine` is located,
+/// and assumes that `raft` is original raftdb directory.
+pub fn check_and_dump_raft_db(new_path: &str, env: Arc<Env>) -> Option<RaftDumpData> {
+    let origin_path = get_raftdb_path(new_path);
+    if !RocksEngine::exists(&origin_path) {
+        return None;
+    }
+    let mut opt = DBOptions::default();
+    opt.set_env(env);
+    let ct_opt = vec![CFOptions::new(CF_DEFAULT, Default::default())];
+    let db = engine_rocks::raw_util::new_engine_opt(&origin_path, opt, ct_opt)
+        .unwrap_or_else(|s| fatal!("failed to create origin raft engine: {}", s));
+    let origin_engine = RocksEngine::from_db(Arc::new(db));
+    let dump_data = dump_data_from_raft_rocksdb(&origin_engine);
+    if dump_data.is_empty() {
+        return None;
+    }
+    Some(dump_data)
+}
+
+/// Recover original raft data into new raft-engine.
+pub fn dump_data_into_raft_engine<E: RaftEngine>(
+    raft_db: &E,
+    new_path: &str,
+    data: Option<RaftDumpData>,
+) {
+    if let Some(data) = data {
+        for (id, (entries, state)) in data {
+            raft_db.put_raft_state(id, &state).unwrap();
+            raft_db.append(id, entries).unwrap();
+        }
+        let origin_path = get_raftdb_path(new_path);
+        std::fs::remove_dir_all(origin_path).expect("Cannot remove raftdb dir");
+    }
+}
+
+fn get_raftdb_path(new_path: &str) -> String {
+    let mut origin_path = PathBuf::from(new_path);
+    origin_path.pop();
+    origin_path.push(CF_RAFT);
+    origin_path.to_str().unwrap().to_owned()
+}
+
+fn dump_data_from_raft_rocksdb(raft_db: &RocksEngine) -> RaftDumpData {
+    let start_key = keys::REGION_RAFT_MIN_KEY;
+    let end_key = keys::REGION_RAFT_MAX_KEY;
+    let mut entries: HashMap<u64, Vec<Entry>> = HashMap::new();
+    let mut raft_states: Vec<(u64, RaftLocalState)> = vec![];
+    raft_db
+        .scan_cf(CF_DEFAULT, start_key, end_key, false, |key, value| {
+            let res = keys::decode_raft_key(key);
+            if res.is_err() {
+                return Ok(true);
+            }
+            let (region_id, suffix) = res.unwrap();
```

Comment:
Why not `let (region_id, suffix) = match ...`?

Reply:
Maybe the number of region will be too large....
We can store all regions in one `LogBatch` and calculate how many entries we have put in and flush it together. Such as 1000 entries perf flush.